README.md
# chemcalculator

[![Documentation Status](https://readthedocs.org/projects/chemcalculator/badge/?version=latest)](https://chemcalculator.readthedocs.io/en/latest/?badge=latest)
[![codecov](https://codecov.io/gh/UBC-MDS/chemcalculator/branch/main/graph/badge.svg?token=pbmgIww2wM)](https://codecov.io/gh/UBC-MDS/chemcalculator)
[![deploy](https://github.com/UBC-MDS/chemcalculator/actions/workflows/ci-cd.yml/badge.svg)](https://github.com/UBC-MDS/chemcalculator/actions/workflows/ci-cd.yml)
## Overview 

chemcalculator is a python package useful for chemistry for the purpose of calculating chemical formula mass in g/mol. The mole allows scientists to calculate the number of elementary entities (usually atoms or molecules) in a certain mass of a given substance. The mass of one mole of a substance is equal to that substance’s molecular weight; as for instance, the mean molecular weight of water is 18.015 atomic mass units (amu), and so one mole of water weighs 18.015 grams. This property simplifies many chemical computations. This python package will be helpful to easily calculate the chemical formula mass, convert moles to grams and vice versa, and lastly calculate the percentage mass for the atomic nature of the elements in chemistry.

This package of basic chemistry calculations is meant to supplement an existing package, [ChemPy](https://github.com/bjodah/chempy), which already handles complex calculations for primarily physical/inorganic/analytical chemistry consisting of, but not limited to, the following:

- Solver for equilibria (including multiphase systems)
- Numerical integration routines for chemical kinetics (ODE solver front-end)
- Integrated rate expressions (and convenience fitting routines)
- Relations in Physical chemistry
- Debye-Hückel expressions
- Arrhenius equation
- Einstein-Smoluchowski equation
- Properties, such as : water density as function of temperature, water permittivity as function of temperature and pressure, and water diffusivity as function of temperature

## Functions

This package contains three functions. Each function will have it's own required and optional arguments.

1. `compute_mass`: Calculate the mass of the atoms or chemical formula for the input chemical formula.
2. `moles_grams_converter`: Convert moles to grams and convert grams to moles.
3. `percent_mass`: Calculate percentage mass for the desired atom or molecule.

## Installation

```bash
$ pip install chemcalculator
```

## Usage

`chemcalculator` can be used as follows:
```bash
from chemcalculator.chemcalculator import compute_mass
compute_mass("H2O")
```
```bash
from chemcalculator.chemcalculator import moles_grams_converter
moles_grams_converter("H2O", 0.05555, "moles")
```
```bash
from chemcalculator.chemcalculator import percent_mass
percent_mass("H2O", "O")
```

## Contributors
### Development Lead

|Contributor Name     | GitHub Username|
|---------------------|-----------|
|Kingslin Lv | [Kingslin0810](https://github.com/Kingslin0810)|
|Joyce Wang      | [jo4356](https://github.com/jo4356)     |
|Allyson Stoll       | [datallurgy](https://github.com/datallurgy) |

We welcome and recognize all contributions. Please find the guide for contribution in [Contributing Document](https://github.com/UBC-MDS/chemcalculator/blob/main/CONTRIBUTING.md).

## License

`chemcalculator` was created by Joyce Wang, Kingslin Lv, Allyson Stoll. It is licensed under the terms of the MIT license.

### Credits

This package was created with Cookiecutter and the UBC-MDS/cookiecutter-ubc-mds project template, modified from the [pyOpenSci/cookiecutter-pyopensci](https://github.com/pyOpenSci/cookiecutter-pyopensci) project template and the [audreyr/cookiecutter-pypackage](https://github.com/audreyr/cookiecutter-pypackage).


CHANGELOG.md
# Changelog

<!--next-version-placeholder-->

## v0.1.12 (2022-02-02)


## v0.1.11 (2022-02-02)


## v0.1.10 (2022-02-01)


## v0.1.9 (2022-02-01)


## v0.1.8 (2022-02-01)


## v0.1.7 (2022-02-01)


## v0.1.6 (2022-01-31)


## v0.1.5 (2022-01-28)


## v0.1.4 (2022-01-28)


## v0.1.3 (2022-01-27)


## v0.1.2 (2022-01-27)


## v0.1.1 (2022-01-27)


## v0.1.0 (11/01/2022)

- First release of `chemcalculator`!

tests/test_ChemCalculator.py
from chemcalculator import chemcalculator


tests/test_chemcalculator.py

from chemcalculator.chemcalculator import compute_mass
from chemcalculator.chemcalculator import moles_grams_converter
from chemcalculator.chemcalculator import percent_mass
import pytest

def test_compute_mass():
    """
    Test the compute_mass function for correct output.
    """
    assert compute_mass('H2O') == pytest.approx(18.013, rel = 1e-3), \
    'compute_mass is not calculated properly (H2O).'
    
    assert compute_mass('C12H22O11') == pytest.approx(342.275, rel = 1e-3), \
    'compute_mass is not calculated properly (C12H22O11).'
    
    assert compute_mass('Al2(SO4)3') == pytest.approx(342.147, rel = 1e-3), \
    'compute_mass is not calculated properly (Al2(SO4)3).'
    
    assert compute_mass('(NH4)HS') == pytest.approx(51.107, rel = 1e-3), \
    'compute_mass is not calculated properly ((NH4)HS).'
    
def test_compute_mass_error():
    """
    Check that error is raised when input is invalid.
    """
    with pytest.raises(TypeError):
        compute_mass(['H2O'])
        
    with pytest.raises(ValueError):
        compute_mass('CuSO4-5H2O')    

    with pytest.raises(ValueError):
        compute_mass('naOH')
    
    with pytest.raises(ValueError):
        compute_mass('(nH4)HS')
        
    with pytest.raises(ValueError):
        compute_mass('NaaaaaaaaOH')
        
def test_moles_grams_converter():
    """
    Test the moles_grams_converter function for correct output
    """
    assert moles_grams_converter("H2O", 18.01528, "grams") == pytest.approx(1, rel=1e-3)
    
    assert moles_grams_converter("H2O", 1, "moles") == pytest.approx(18.015, rel=1e-3)

def test_moles_grams_converter_error():
    """
    Check that error is raised when input is invalid.
    """
    with pytest.raises(TypeError):
        moles_grams_converter('H2O', 1, 'tons')
        
    with pytest.raises(TypeError):
        moles_grams_converter(['H2O'], 1, 'moles')
        
    with pytest.raises(ValueError):
        moles_grams_converter('CuSO4-5H2O', 1, 'moles')    

    with pytest.raises(ValueError):
        moles_grams_converter('naOH', 1, 'moles')
    
    with pytest.raises(ValueError):
        moles_grams_converter('(nH4)HS', 1, 'moles')
        
    with pytest.raises(ValueError):
        moles_grams_converter('NaaaaaaaaOH', 1, 'moles')

def test_percent_mass():
    """
    Test the percent_mass function for correct output
    """
    assert percent_mass("H2O", "O") == pytest.approx(88.819, rel=1e-3), \
    "precent_mass is not calculated correctly (H2O, O)"
    
    assert percent_mass("H2O", "H") == pytest.approx(5.59, rel=1e-3), \
    "precent_mass is not calculated correctly (H2O, H)"

    assert percent_mass("H2O", "H2") == pytest.approx(11.181, rel=1e-3), \
    "precent_mass is not calculated correctly (H2O, H2)"

    assert percent_mass("NaOH", "OH") == pytest.approx(42.519, rel=1e-3), \
    "precent_mass is not calculated correctly (NaOH, OH)"

    assert percent_mass("H2SO4", "SO4") == pytest.approx(97.946, rel=1e-3), \
    "precent_mass is not calculated correctly (H2SO4, SO4)"

    assert percent_mass("H2SO4", "HSO4") == pytest.approx(98.973, rel=1e-3), \
    "precent_mass is not calculated correctly (H2SO4, HSO4)"

    assert percent_mass("Na(H2SO4)3", "NaH2SO4") == pytest.approx(38.165, rel=1e-3), \
    "precent_mass is not calculated correctly (Na(H2SO4)3, NaH2SO4)"


def test_percent_mass_error():
    """Check that error is raised when input is incorrect"""
    with pytest.raises(ValueError):
        percent_mass("H2SO4", "NaSO4")
    with pytest.raises(ValueError):
        percent_mass("H2SO4", "Al")
    with pytest.raises(ValueError):
        percent_mass("H2SO4", "H2SO4C5")
    with pytest.raises(ValueError):
        percent_mass("H2SO4", "H3")
    with pytest.raises(ValueError):
        percent_mass("H2SO4", "H3S5")
    with pytest.raises(ValueError):
        percent_mass(".", "H2O")
    with pytest.raises(ValueError):
        percent_mass("NaOH", "sss")
    with pytest.raises(TypeError):
        percent_mass(3, "O")

        


src/chemcalculator/chemcalculator.py

import pandas as pd
from chemcalculator.datasets import get_periodic_table
from chemcalculator.helper import __chemical_elements, __check_chemical_format


def compute_mass(chemical):
    """Computes the molar mass of the given chemical compound.

    Parameters
    ----------
    chemical : string
        The molecular formula of the given chemical compound given as a string.

    Returns
    -------
    float
        The molar mass of the given formula in g/mol.

    Raises
    ------
    TypeError
        Entered value is not a string.
    ValueError
        String contains characters that are not allowed.
    ValueError
        String or subcomponent starts with lowercase letter.
    ValueError
        Chemical compound contains element not in periodic table.

    Examples
    --------
    >>> compute_mass("H2O")
    18.01528

    >>> compute_mass("C12H22O11")
    342.3

    >>> compute_mass("Al2(SO4)3")
    342.15
    
    >>> compute_mass("(NH4)HS")
    51.107
    """
    __check_chemical_format(chemical)
    
    raw_elements = __chemical_elements(chemical)

    periodic_table_mass = get_periodic_table()
    
    if set(raw_elements.keys()).issubset(periodic_table_mass): pass
    else:
        raise ValueError('Chemical compound contains element not in periodic table.')
    
    # count and sum up all individual elements
    df = pd.DataFrame.from_dict(raw_elements,
                                orient='index',
                                columns = ["Counts"]).reset_index()
    df = df.rename(columns={"index": "Element"})
    
    # map atomic masses from periodic table and multiply by atomic counts
    df['Mass'] = df['Element'].map(periodic_table_mass) * df['Counts']
    
    # sum and return mass
    return df['Mass'].sum()

def moles_grams_converter(formula, mass, convert_from):
    """
    Converts moles to grams or grams to moles

    Parameters
    ----------
    formula : string
        the checmical formula for the conversion
    mass : float
        the mass of molecule that needs to be converted (grams or moles)
    converter : string
        indicates to convert either "moles" or "grams"

    Returns
    -------
    float
        mass that is converted to either moles or grams

    Examples
    --------
    >>> moles_grams_converter("H2O", 0.05555, "moles")
    1.000

    >>> moles_grams_converter("H2O", 18.01528, "grams")
    1.000
    """
    grams_per_mole = compute_mass(formula)
    if convert_from == "grams":
        result = mass / grams_per_mole
    elif convert_from == "moles":
        result = mass * grams_per_mole
    else:
        raise TypeError("Please enter either 'grams' or 'moles'")

    return round(result, 3)


def percent_mass(compound, element):
    """
    Calculates the percentage mass of an element (or compound) in a compound

    Parameters
    ----------
    compound : string
        the chemical formula of the full compound
    element : string
        the chemical formula of the element or compound of interest

    Returns
    -------
    float
        the percentage mass of the element or compound of interest
    
    Examples
    --------
    >>> percent_mass("H2O", "O")
    88.819
    
    >>> percent_mass("H2O", "H")
    5.59

    >>> percent_mass("H2O", "H2")
    11.181

    >>> percent_mass("NaOH", "OH")
    42.519
    """
    
    __check_chemical_format(compound)
    __check_chemical_format(element)

    perc_mass = 0
    compound_count = __chemical_elements(compound)
    element_count = __chemical_elements(element)

    if set(element_count).issubset(compound_count): 
        for elem in element_count:
            if element_count[elem] <= compound_count[elem]:
                perc_mass = round(compute_mass(element)/compute_mass(compound)*100, 3)
            else:
                raise ValueError("There cannot be more counts of elements in the sub-compound compared to the larger compound")
    else:
        raise ValueError("Please make sure the sub-compound is part of the larger compound")

    print(f"The percentage mass of {element} in {compound} is: {perc_mass} %")
    return perc_mass


